### DISCUSSION TITLE
Response times of public "what is my ip" services.

### DISCUSSION THREAD

**Discussion by fdcastel** [troglobit/inadyn#456] [Show and tell]
For reference.

I made [a simple Powershell script](https://gist.github.com/fdcastel/b5250cb47b86a53f394443cdaf44b940) to test the current Inadyn default service (https://ifconfig.me/ip) against others similar services from Cloudflare, Amazon and Google.

Results below:

- From **North America/East** (OVH / Beauharnois, Canada):

    ```
    Name                           Value
    ----                           -----
    https://1.1.1.1/cdn-cgi/trace  10,0242
    https://checkip.amazonaws.com/ 14,9672
    https://ifconfig.me/ip         34,5791
    https://domains.google.com/... 66,07765
    ```

- From **Europe/West** (Hetzner / Falkenstein, Germany):
    ```
    Name                           Value
    ----                           -----
    https://1.1.1.1/cdn-cgi/trace  25.7611625
    https://checkip.amazonaws.com/ 124.2826
    https://ifconfig.me/ip         134.1107625
    https://domains.google.com/ch… 263.3470375
    ```

- From **South America** (Vivo / São Paulo, Brazil):

    ```
    Name                           Value                                                                                                                                                                                      
    ----                           -----                                                                                                                                                                                      
    https://1.1.1.1/cdn-cgi/trace  16,6689                                                                                                                                                                                    
    https://ifconfig.me/ip         159,5154375                                                                                                                                                                                
    https://checkip.amazonaws.com/ 161,71955                                                                                                                                                                                  
    https://domains.google.com/... 792,0165125
    ```

Anyone is invited to run the script and publish here the results for other regions.

But I am almost sure Cloudflare would always be the winner due to its distributed nature. There is always a `1.1.1.1` server very close to whenever you are in the globe.

_P.S.: This is not a call for a change, just a little toy measurement. The actual process of update a DNS record made by Inadyn requires several other API calls which would quickly eclipse any gains derived from this change alone._

_That said, if anyone wish to change the default service to Cloudflare... I surely have nothing against it._ 😉

**Comment by troglobit** [comment]
I would not mind changing the default for the fallback checkip service.  My concern is finding a service that's reliable and does not change it's API all the time.  Reliability was a real concern before we changed to the one we use now, and prior to that we relied on Dyn.com as fallback, but they started throttling their replies.  Since Inadyn is part of many distributions, and embedded systems, that run for years, often unmodified, it caused a few bug reports that were impossible to address.  So I've considered making this a .conf file option actually.

Only little snag here is that the Cloudflare output cannot be immediately parsed by Inadyn as-is, so some extra logic has to be bolted on, e.g. a script filter or similar.

**Reply by fdcastel** [reply]
> Only little snag here is that the Cloudflare output cannot be immediately parsed by Inadyn as-is, so some extra logic has to be bolted on, e.g. a script filter or similar.

Yes. Clouflare answer is something like

```
fl=299f4
h=1.1.1.1
ip=XXX.XXX.215.113
ts=1695895166.854
...
```

Which requires some parsing to extract the `ip` part.

However... I didn't get a closer look into this part of the code. But the fact that I found several providers using `/cdn-cgi/trace` made me think that this was already a solved problem.

**Reply by troglobit** [reply]
Interesting, that I'll have to investigate. It's possible there's something like a strstr() for "ip=" somewhere in the reptile brain of the code

**Reply by troglobit** [reply]
I've looked at the code now, can't really see the output from /cdn-cgi/trace being parsed safely.  The code is very opportunistic, the first thing that looks like an IP address is accepted ... so without verifying it, I'm afraid it would pick 1.1.1.1 from the above output instead of XXX.XXX.215.113

**Reply by fdcastel** [reply]
Ugh!

Time for a `git blame`... 😅

**Reply by troglobit** [reply]
Water under the bridge ... don't look, it probably ends up being something I did :-D

I really think all we need is a very simple filter to complete this.  Like, "only use line starting with" or "line matching substring".  Should be fairly simple:

    checkip-filter = "match:ip="
    checkip-filter = "substr:myaddress"

It opens up for crazy pull-requests though, so we might end up with a complete `awk` or `sed` in there.  Just an idea anyway.

**Comment by fdcastel** [comment]
> My concern is finding a service that's reliable and does not change it's API all the time. Reliability was a real concern before 

Fully understood. 

That said -- and as you already may have noticed -- I have a tendency to bet my chips on Cloudflare 😅.

I can only imagine the problems you had in the past. But if we take a step back and look at the larger picture... Practically ALL of this "dynamic dns services" are _small fish_. That's why I looked only for the larger ones in this test. And, frankly, the differences between the regions took me by surprise.

If I had to choose just one of all Inadyn providers currently implemented I... Well... You know the answer. 😉 

Anyway: having options are good. The current default choice is good. The benefit to change it is very small. 

As I said, this exercise was just to get some numbers and put things in perspective.
